internal/runtime/storeforward.go

```diff
@@ -0,0 +1,188 @@
+				sf.retryStoredData(serviceKey, config, edgeXClients)
+			}
+		}
+
+		edgeXClients.LoggingClient.Info("StoreAndForward Retry Loop exited")
+	}()
+}
+
+func (sf *storeForwardInfo) storeForLaterRetry(payload []byte, edgexcontext *appcontext.Context, pipelinePosition int) {
+	item := contracts.NewStoredObject(sf.runtime.ServiceKey, payload, pipelinePosition, sf.pipelineHash)
+	item.CorrelationID = edgexcontext.CorrelationID
+	item.EventID = edgexcontext.EventID
+	item.EventChecksum = edgexcontext.EventChecksum
+
+	edgexcontext.LoggingClient.Trace("Storing data for later retry", clients.CorrelationHeader, edgexcontext.CorrelationID)
+
+	if !edgexcontext.Configuration.Writable.StoreAndForward.Enabled {
+		edgexcontext.LoggingClient.Error("Failed to store item for later retry", "error", "StoreAndForward not enabled", clients.CorrelationHeader, item.CorrelationID)
+		return
+	}
+
+	if _, err := sf.storeClient.Store(item); err != nil {
+		edgexcontext.LoggingClient.Error("Failed to store item for later retry", "error", err, clients.CorrelationHeader, item.CorrelationID)
+	}
+}
+
+func (sf *storeForwardInfo) retryStoredData(serviceKey string, config *common.ConfigurationStruct, edgeXClients common.EdgeXClients) {
+	items, err := sf.storeClient.RetrieveFromStore(serviceKey)
+	if err != nil {
+		edgeXClients.LoggingClient.Error("Unable to load store and forward items from DB", "error", err)
+		return
+	}
+
+	edgeXClients.LoggingClient.Debug(fmt.Sprintf(" %d stored data items found for retrying", len(items)))
+
+	if len(items) > 0 {
+		itemsToRemove, itemsToUpdate := sf.processRetryItems(items, config, edgeXClients)
+
+		edgeXClients.LoggingClient.Debug(fmt.Sprintf(" %d stored data items will be removed post retry", len(itemsToRemove)))
+		edgeXClients.LoggingClient.Debug(fmt.Sprintf(" %d stored data items will be update post retry", len(itemsToUpdate)))
+
+		for _, item := range itemsToRemove {
+			if err := sf.storeClient.RemoveFromStore(item); err != nil {
+				edgeXClients.LoggingClient.Error("Unable to remove stored data item from DB", "error", err, "objectID", item.ID, clients.CorrelationHeader, item.CorrelationID)
+			}
+		}
+
+		for _, item := range itemsToUpdate {
+			if err := sf.storeClient.Update(item); err != nil {
+				edgeXClients.LoggingClient.Error("Unable to update stored data item in DB", "error", err, "objectID", item.ID, clients.CorrelationHeader, item.CorrelationID)
+			}
+		}
+	}
+}
+
+func (sf *storeForwardInfo) processRetryItems(items []contracts.StoredObject, config *common.ConfigurationStruct, edgeXClients common.EdgeXClients) ([]contracts.StoredObject, []contracts.StoredObject) {
+	var itemsToRemove []contracts.StoredObject
+	var itemsToUpdate []contracts.StoredObject
+
+	for _, item := range items {
+		validVersion := item.Version == sf.calculatePipelineHash()
+
+		if validVersion {
+			success := sf.retryExportFunction(item, config, edgeXClients)
+			if !success {
```

Comment:
Why not combine this and the prior line into a single if statement?

Reply:
Nitpick - no change in code.

Reply:
Ahh, yep, missed that one... ;-)